pennylane/shadows/transforms.py

```diff
@@ -0,0 +1,212 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classical shadow transforms"""
+
+from itertools import product
+from functools import reduce, wraps
+import warnings
+
+import pennylane as qml
+import pennylane.numpy as np
+
+
+@qml.batch_transform
+def __replace_obs(tape, obs, *args, **kwargs):
+    """
+    Tape transform to replace the measurement processes with the given one
+    """
+    with qml.tape.QuantumTape() as new_tape:
+        # queue everything from the old tape except the measurement processes
+        for op in tape.operations:
+            qml.apply(op)
+
+        # queue the new observable
+        obs(*args, **kwargs)
+
+    def processing_fn(res):
+        return qml.math.squeeze(qml.math.stack(res))
+
+    return [new_tape], processing_fn
+
+
+def shadow_expval(H, k=1):
+    """Transform a QNode returning a classical shadow into one that returns
+    the approximate expectation values in a differentiable manner.
```

Comment:
Would be nice to check that the measurement is indeed `qml.classical_shadow` and raise an error otherwise. Right now you can use this transform with any measurement, which is not very intuitive.

Reply:
On another note, what is the difference between using this transform and directly using the measurement `qml.shadow_expval`? It feels like it is the exact same thing.

If it is I don't see the reason for having this transform (or maybe we could remove the `qml.shadow_expval` measurement and move all the functionality here?).